### PR TITLE
General SA Improvements & Low-Hanging Fruit

### PR DESCRIPTION
**Pull request by gsteel**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| QA            | yes

### Description

General reduction is psalm baseline - mostly corrections to documented parameter/return types, addition of array shapes for various options. Descendants of `AbstractUnicode` benefit from some extra code removal

### CODE REVIEW

File: src/Boolean.php [hunk 1]
```diff
@@ -19,6 +19,14 @@
 use function sprintf;
 use function strtolower;
 
+/**
+ * @psalm-type Options = array{
+ *     type: int,
```
Comment (Ocramius) [inline src/Boolean.php]: Is the type `one-of<self::TYPE_*>`?

File: src/ToNull.php [hunk 2]
```diff
@@ -16,6 +16,12 @@
 use function iterator_to_array;
 use function sprintf;
 
+/**
+ * @psalm-type Options = array{
+ *     type: int,
```
Comment (Ocramius) [inline src/ToNull.php]: Can we restrict this integer further?

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Boolean.php [hunk 3]
```diff
@@ -19,6 +19,14 @@
 use function sprintf;
 use function strtolower;
 
+/**
+ * @psalm-type Options = array{
+ *     type: int-mask-of<self::TYPE_*>,
```
Comment (Ocramius) [inline src/Boolean.php:24]: Wait, is it a one-of, or an int-mask-of? Confusing across the file 🤔

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Boolean.php (hunk 3 above)
Comment (gsteel) [reply src/Boolean.php:24]: Internally, it's in-mask. As a parameter, it's sometimes a string equivalent

Comment (Ocramius) [reply src/Boolean.php:24]: Thanks for clarifying 👍

**Review by Ocramius — APPROVED** [review]
Most excellent, thanks @gsteel!

**Pull request merged**